dbms/src/Client/PerformanceTest.cpp

```diff
@@ -0,0 +1,1556 @@
+        }
+
+        size_t max_bytes_to_read = stop_criterions[statistic_index].bytes_read_uncompressed.value;
+        if (max_bytes_to_read && statistics[statistic_index].bytes_read >= max_bytes_to_read)
+        {
+            incFulfilledCriterions(statistic_index, bytes_read_uncompressed);
+        }
+
+        if (UInt64 max_timeout_ms = stop_criterions[statistic_index].timeout_ms.value)
+        {
+            /// cast nanoseconds to ms
+            if ((statistics[statistic_index].watch.elapsed() / (1000 * 1000)) > max_timeout_ms)
+            {
+                incFulfilledCriterions(statistic_index, timeout_ms);
+            }
+        }
+
+        size_t min_time_not_changing_for_ms = stop_criterions[statistic_index].min_time_not_changing_for_ms.value;
+        if (min_time_not_changing_for_ms)
+        {
+            size_t min_time_did_not_change_for = statistics[statistic_index].min_time_watch.elapsed() / (1000 * 1000);
+
+            if (min_time_did_not_change_for >= min_time_not_changing_for_ms)
+            {
+                incFulfilledCriterions(statistic_index, min_time_not_changing_for_ms);
+            }
+        }
+
+        size_t max_speed_not_changing_for_ms = stop_criterions[statistic_index].max_speed_not_changing_for_ms.value;
+        if (max_speed_not_changing_for_ms)
+        {
+            UInt64 speed_not_changing_time = statistics[statistic_index].max_rows_speed_watch.elapsed() / (1000 * 1000);
+            if (speed_not_changing_time >= max_speed_not_changing_for_ms)
+            {
+                incFulfilledCriterions(statistic_index, max_speed_not_changing_for_ms);
+            }
+        }
+
+        size_t average_speed_not_changing_for_ms = stop_criterions[statistic_index].average_speed_not_changing_for_ms.value;
+        if (average_speed_not_changing_for_ms)
+        {
+            UInt64 speed_not_changing_time = statistics[statistic_index].avg_rows_speed_watch.elapsed() / (1000 * 1000);
+            if (speed_not_changing_time >= average_speed_not_changing_for_ms)
+            {
+                incFulfilledCriterions(statistic_index, average_speed_not_changing_for_ms);
+            }
+        }
+
+        if (stop_criterions[statistic_index].number_of_initialized_min
+            && (stop_criterions[statistic_index].fulfilled_criterions_min >= stop_criterions[statistic_index].number_of_initialized_min))
+        {
+            /// All 'min' criterions are fulfilled
+            stream.cancel();
+        }
+
+        if (stop_criterions[statistic_index].number_of_initialized_max && stop_criterions[statistic_index].fulfilled_criterions_max)
+        {
+            /// Some 'max' criterions are fulfilled
+            stream.cancel();
+        }
+
+        if (interrupt_listener.check())
+        {
+            gotSIGINT = true;
+            stream.cancel();
+        }
+    }
+
+    void constructSubstitutions(AbstractConfig & substitutions_view, StringToVector & substitutions)
```

Comment:
Maybe `const AbstractConfig & substitutions_view`?
And a comment would be helpful.